initial commit to pcb_alignment

Keeginator91 opened the pull request:
There was a discrepancy between what Arduino pins were being used for which MOSFETS. This aims to fix that. This was done by creating macros for each Arduino pin being used and naming them with their respective schematic connection ID. These pin ID macros are then assigned to the their respective `CHG_FET` or `OUT_FET`. This allowed for the issue to be resolved, readability to be maintained, and the least amount of code changes.

Translation table of what MOSFET is control by what gate driver, gate driver pin, and Arduino pin and how that is used in pin assignments in the code.
![image](https://github.com/Keeginator91/Capstone_Solar_Cart/assets/103848741/10bd3493-c4e4-4075-bcc2-a7281dfb8824)


On top of this I moved any constant arrays above `void setup()` and made them `const` variables since they shouldn't be changed once we are running.

These are the results of an in-browser compiler I use to make sure that the logic and assignments of the assignment functions are correct.
```
Battery 0 configuration
    is_charging: 0
    adc_pin_assignment: 54
    OUT_FETS: 3, 11, 23, 28, 30, 
    CHARGE_FETS: 2, 5, 
Battery 1 configuration
    is_charging: 0
    adc_pin_assignment: 55
    OUT_FETS: 4, 6, 23, 28, 30, 
    CHARGE_FETS: 8, 9, 
Battery 2 configuration
    is_charging: 0
    adc_pin_assignment: 56
    OUT_FETS: 4, 7, 10, 28, 30, 
    CHARGE_FETS: 12, 22, 
Battery 3 configuration
    is_charging: 0
    adc_pin_assignment: 57
    OUT_FETS: 4, 7, 11, 13, 30, 
    CHARGE_FETS: 24, 25, 
Battery 4 configuration
    is_charging: 0
    adc_pin_assignment: 58
    OUT_FETS: 4, 7, 11, 23, 26, 
    CHARGE_FETS: 27, 29, 
```

Keeginator91 committed:
initial commit to pcb_alignment

There was a discrepancy between what Arduino pins where being used for which MOSFETS. This aims to fix that.

Keeginator91 committed:
static int -> const int

mdesantis30 commented on Array_Control/Array_Control.ino:
```diff
@@ -23,17 +20,18 @@
-#define NUM_FETS  5  //Number of FETS used to bypass a battery
-#define FET_ARRAY_LEN 11 //total number of FETS used
+#define NUM_FETS  5  //Number of FETS used per a battery
+#define NUM_CHG_FETS  10 //total number of charging fets   
```
number of charge FETs should be the same as the number as output FETs

Keeginator91 replied:
There is one more output FET than charging FETs. I just made a commit changing line #25 of .ino to `NUM_OUT_FETS` to make this clearer.

TCecelya reviewed (approved):
This is good, we do not want to accidentally change these arrays even though our code does not intentionally do so.

TCecelya reviewed (approved):
I see where you are going with this change - I was a little confused looking at Array_Control.ino before I saw how you defined those macros in Array_Control.h. I don't notice any glaring errors but this is obviously without performing any testing of the code (although I do trust that it compiles through your online compiler). TC.

mdesantis30 commented on Array_Control/Array_Control.ino:
```diff
@@ -42,10 +40,10 @@ bool DEBUG = true;
-#define MOSFET_FALL_TIME 0.000000115
+#define MOSFET_TURN_ON   0.000000021f 
+#define MOSFET_RISE_TIME 0.000000120f 
+#define MOSFET_TURN_OFF  0.000000180f 
```
why the "f"?

Keeginator91 replied:
the `f` makes it a floating point instead of a double. Allows for better accuracy.

Keeginator91 committed:
update .ino

changed FET_ARRAY_LEN to NUM_OUT_FETS for improved readability/clarity

Keeginator91 committed:
Moved macros and rename

moved macros with defined integer values to .h do remove magic number in the battery structure declaration.

Also renamed NUM_FETS to NUM_OUT_FETS_PER_BATT because thats actually what it is.

mdesantis30 commented:
well done, I like these changes. It all makes sense.

TCecelya commented on Array_Control/Array_Control.ino:
```diff
@@ -22,30 +19,6 @@
-#define MOSFET_RISE_TIME 0.000000120 
-#define MOSFET_TURN_OFF  0.000000180 
-#define MOSFET_FALL_TIME 0.000000115
 #define MOSFET_ON_DELAY  (MOSFET_TURN_ON  + MOSFET_RISE_TIME)
```
Should this pre-processor directive/macro definition for MOSFET on time/reibe moved to Array_Control.h as well?

Keeginator91 replied:
fixed in latest commit

TCecelya commented on Array_Control/Array_Control.ino:
```diff
@@ -148,48 +136,46 @@ battery batts_array[NUM_BATTS]; // {batt_0, batt_1, ...}
+            //printf("        CHARGE_FETS: ");
+            Serial.print("      CHARGE_FETS: ");
+
+            for(int ii = 0; ii < 2; ii++){
```
I would change the condition "ii < 2" to "ii < NUM_CHARGE_FETS_PER_BATT" to keep it more general. The same applies above on line 163 for the conditional for output FETs.

Keeginator91 replied:
fixed in latest commit

TCecelya reviewed (changes requested):
I left comments, but overall the changes are good (I am assuming I am reviewing the readability changes you made after the initial pull request). TC.

Keeginator91 committed:
moved Macros and other updates

moved the rest of the non macro-assigned-to-macros to the .h file.
Updated some small formatting stuff and comments in .ino and .h.

TCecelya committed:
Added changes to accommodate charging MOSFETs (ensuring they are properly turned on and turned off). Adjusted logic for determining which battery is the lowest - It does not depending on if the battery is not charging, because this flag is automatically set to false in the setup function and is never set to true.

TCecelya committed:
Removed unnecessary serial debug/testing code from void loop(). Pushing updated changes made for testing to remote repository. TC

Keeginator91 committed:
added updated flow diagrams of program functions

Keeginator91 committed:
Add updated images apparently
